src/game/client/heightmap.cpp

```diff
@@ -0,0 +1,378 @@
+        ret = 0;
+    }
+
+    if (ret >= m_x - 1) {
+        return m_x - 1;
+    }
+
+    return ret;
+}
+
+int HeightMapRenderObjClass::Get_Y_With_Origin(int y)
+{
+    int ret = y - m_originY;
+
+    if (ret < 0) {
+        ret = ret + m_y - 1;
+    }
+
+    if (ret >= m_y - 1) {
+        ret -= m_y - 1;
+    }
+
+    if (ret < 0) {
+        ret = 0;
+    }
+
+    if (ret >= m_y - 1) {
+        return m_y - 1;
+    }
+
+    return ret;
+}
+
+int HeightMapRenderObjClass::Update_VB(DX8VertexBufferClass *vb,
+    char *data,
+    int x0,
+    int y0,
+    int x1,
+    int y1,
+    int origin_x,
+    int origin_y,
+    WorldHeightMap *map,
+    RefMultiListIterator<RenderObjClass> *lights)
+{
+    // heightmap TODO
+#ifdef GAME_DLL
+    return Call_Method<int,
+        HeightMapRenderObjClass,
+        DX8VertexBufferClass *,
+        char *,
+        int,
+        int,
+        int,
+        int,
+        int,
+        int,
+        WorldHeightMap *,
+        RefMultiListIterator<RenderObjClass> *>(
+        PICK_ADDRESS(0x00799610, 0x0060622E), this, vb, data, x0, y0, x1, y1, origin_x, origin_y, map, lights);
+#else
+    return 0;
+#endif
+}
+
+int HeightMapRenderObjClass::Update_VB_For_Light(DX8VertexBufferClass *vb,
+    char *data,
+    int x0,
+    int y0,
+    int x1,
+    int y1,
+    int origin_x,
+    int origin_y,
+    W3DDynamicLight **lights,
+    int num_lights)
+{
+    // heightmap TODO
+#ifdef GAME_DLL
+    return Call_Method<int,
+        HeightMapRenderObjClass,
+        DX8VertexBufferClass *,
+        char *,
+        int,
+        int,
+        int,
+        int,
+        int,
+        int,
+        W3DDynamicLight **,
+        int>(PICK_ADDRESS(0x0079ACA0, 0x00607629), this, vb, data, x0, y0, x1, y1, origin_x, origin_y, lights, num_lights);
+#else
+    return 0;
+#endif
+}
+
+int HeightMapRenderObjClass::Update_VB_For_Light_Optimized(DX8VertexBufferClass *vb,
+    char *data,
+    int x0,
+    int y0,
+    int x1,
+    int y1,
+    int origin_x,
+    int origin_y,
+    W3DDynamicLight **lights,
+    int num_lights)
+{
+    // heightmap TODO
+#ifdef GAME_DLL
+    return Call_Method<int,
+        HeightMapRenderObjClass,
+        DX8VertexBufferClass *,
+        char *,
+        int,
+        int,
+        int,
+        int,
+        int,
+        int,
+        W3DDynamicLight **,
+        int>(PICK_ADDRESS(0x0, 0x006080DB),
+        this,
+        vb,
+        data,
+        x0,
+        y0,
+        x1,
+        y1,
+        origin_x,
+        origin_y,
+        lights,
+        num_lights); // does not exist in game exe
+#else
+    return 0;
+#endif
+}
+
+void HeightMapRenderObjClass::Render_Terrain_Pass(CameraClass *camera)
+{
+#ifdef BUILD_WITH_D3D8
+    Matrix3D tm(true);
+    DX8Wrapper::Set_Transform(D3DTS_WORLD, tm);
+    DX8Wrapper::Set_Index_Buffer(m_indexBuffer, 0);
+
+    for (int i = 0; i < m_numVBTilesY; i++) {
+        for (int j = 0; j < m_numVBTilesX; j++) {
+            static int count = 0;
```

Comment:
needs to be at top else it sets it every loop

Reply:
fixed.